File: 503-next-greater-element-ii/next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
         map<int,int>m;
        stack<int>st;
        for(int i=0;i<2*nums.size();i++){
            int j=i%nums.size();
            while(!st.empty() && nums[st.top()]<nums[j]){
              if(m.find(st.top())==m.end()) m[st.top()]=nums[j];
                st.pop();
            }
            st.push(j);
        }
        while(!st.empty()){
         if(m.find(st.top())==m.end())   m[st.top()]=-1;
            st.pop();
        }

        for(int i=0;i<nums.size();i++){
            nums[i]=m[i];
        }
        return nums;
    }
};

File: 503-next-greater-element-ii/next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n=nums.length;
        Stack<Integer>st=new Stack<>();
        int []ans=new int[n];
        for(int i=2*n-1;i>=0;i--){
            if(i<n){
               while(!st.isEmpty() && st.peek()<=nums[i]){
                st.pop();
               } 
                ans[i]=!st.isEmpty()?st.peek():-1;
            }

            st.push(nums[i%n]);
        }

        return ans;
    }
}